File: day23/ocaml/fun.ml



(* declare some mutable variables

   let a = ref 0 ;;
   a is now a reference ? to Int
   to get value of a we dereference using ! operator
   !a
   if (!a != 0) then a is not 0
                else a is zero ...

 *)
let a = ref 0;;
let b = ref 0 ;;
let c = ref 0 ;;
let d = ref 0 ;;

(* some procedures *)
let reset = fun () -> a := 0 ; b := 0; c:= 0; d:= 0 ;;

let rec v0 () = a := 1 ; v1 ()
and v1 () = b := 1 ; v2()
and v2 () = d := 26 ; v3()
and v3 () = if (!c != 0) then v5() else v4()
and v4 () = v9()
and v5 () = c := 7 ; v6()
and v6 () = d := !d + 1 ; v7()
and v7 () = c := !c - 1 ; v8()
and v8 () = if (!c != 0) then v6() else v9() 
and v9 () = c := !a ; v10()
and v10 () = a := !a + 1 ; v11() 
and v11 () = b := !b - 1 ; v12() 
and v12 () = if (!b != 0) then v10() else v13() 
and v13 () = b := !c ; v14() 
and v14 () = d := !d - 1 ; v15() 
and v15 () = if (!d != 0) then v9 () else v16()
and v16 () = c := 18 ; v17 ()
and v17 () = d := 11 ; v18()
and v18 () = a := !a + 1 ; v19() 
and v19 () = d := !d - 1 ; v20()
and v20 () = if (!d != 0) then v18() else v21()
and v21 () = c := !c - 1 ; v22()
and v22 () = if (!c != 0) then v17() else vDone()
and vDone () = ()

let run () =  reset();
              v0();
              Format.printf "--------------------------\n" ;
              Format.printf "a = %d\n" !a ;  
              Format.printf "b = %d\n" !b ; 
              Format.printf "c = %d\n" !c ;
              Format.printf "d = %d\n" !d ;
              Format.printf "--------------------------\n" 


let run2 () =  reset() ;
               c := 1 ;
               v0();
               Format.printf "--------------------------\n" ;
               Format.printf "a = %d\n" !a ;
               Format.printf "b = %d\n" !b;
               Format.printf "c = %d\n" !c;
               Format.printf "d = %d\n" !d;
               Format.printf "--------------------------\n" 



let _ = run();;
let _ = run2();;

(* let partA () = run() ; !a ;; *)
(*  *)
(* let partB () = run2() ; !a ;; *)

File: day23/lua/fun.lua


--[[
   this is a lua solution comment
   keyword landmines
    and       break     do        else      elseif
    end       false     for       function  if
    in        local     nil       not       or
    repeat    return    then      true      until
   while

   function !=(x,y)
   return not( x == y)
   end
   
--]]

a = 0
b = 0
c = 0
d = 0

function reset()
   a = 0
   b = 0
   c = 0
   d = 0
end

function v0 ()
   a = 1
   return v1()
end

function v1 ()
   b = 1
   return v2()
end

function v2()
   d = 26
   return v3()
end

function v3()
   if not (c == 0)
   then return v5()
   else return v4()
   end
end
   
function v4()
  return v9()
end

function v5()
   c = 7
   return v6()
end
function v6()
   d = d + 1
   return v7()
end

function v7()
   c = c -1
   return v8()
end

function v8()
   if not(c == 0)
   then return v6()
   else return v9()
   end
end

function v9()
   c = a
   return v10()
end

function v10()
   a = a + 1
   return v11()
end

function v11()
   b = b - 1
   return v12()
end

function v12()
   if not (b == 0)
   then return v10()
   else return v13()
   end
end

function v13()
   b = c
   return v14()
end

function v14()
   d = d - 1
   return v15()
end

function v15()
   if not(d == 0)
   then return v9()
   else return v16()
   end
end

function v16()
   c = 18
   return v17()
end

function v17()
   d = 11
   return v18()
end

function v18()
   a = a + 1
   return v19()
end

function v19()
   d = d - 1
   return v20()
end

function v20()
   if not(d == 0)
   then return v18()
   else return v21()
   end
end

function v21()
   c = c - 1
   return v22()
end

function v22()
   if not(c ==0)
   then return v17()
   else return vDone()
   end   
end

function vDone()
   return true
end


function run()
   reset()
   v0()
   print("\nRun 1 \n")
   print("a = " , a)
   print("b = " , b)
   print("c = " , c)
   print("d = " , d)   
end

function run2()
   reset()
   c = 1
   v0()
   print("\nRun 2 with c = 1 \n")
   print("a = " , a)
   print("b = " , b)
   print("c = " , c)
   print("d = " , d)   
end


run()
run2()

--[[

   in order for tail call recursion to take place
   lua needs explicit return statement


Run 1 

a = 	318009
b = 	196418
c = 	0
d = 	0

Run 2 with c = 1 

a = 	9227663
b = 	5702887
c = 	0
d = 	0

real	0m0.402s
user	0m0.401s
sys	0m0.001s

   v = {}
   v[0] = v0
   v[1] = v1
   v[2] = v2 ... etc..
   
   
--]]

